File: MatlabCodes/Chapter 04/Fig04_10.m
function LocalVolHagan
close all;clf;

% For the SABR model we take beta =1 and rho =0 (simpilification)

beta   = 1;
rho    = 0;

% Other model parameters

volvol = 0.2;
s0     = 1;
T      = 10;
r      = 0.05;
alpha  = 0.2;
f_0    = s0*exp(r*T);

% Monte Carlo settings

NoOfPaths = 25000;
NoOfSteps = 100*T;
randn('seed',2)

% We define the market to be driven by Hagan's SABR formula
% Based on this formula we derive local volatility/variance 

sigma     = @(x,t) HaganImpliedVolatility(x,t,f_0,alpha,beta,rho,volvol);
sigma_ATM = HaganImpliedVolatility(f_0,T,f_0,alpha,beta,rho,volvol);

% Local variance based on the Hagan's SABR formula

sigmalv2  = LocalVarianceBasedOnSABR(s0,f_0,r,alpha,beta,rho,volvol);

dt        = T/NoOfSteps;
Z         = random('normal',0,1,[NoOfPaths,NoOfSteps]);
S         = zeros(NoOfPaths,NoOfSteps+1);

S(:,1)    = s0;
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps

    % This condition is necessary as for t=0 we cannot compute implied
    % volatilities

    if time(i)==0
        time(i)=0.0001;
    end
    sprintf('current time is %3.3f',time(i))

    % Standarize Normal(0,1)

    Z(:,i)=(Z(:,i)-mean(Z(:,i)))/std(Z(:,i));
    
    % Compute local volatility

    sig=real(sigmalv2(S(:,i),time(i)));

    % Because of discretizations we may encounter negative variance which
    % needs to be set to 0.

    sig=max(sig,0);
    sigmaLV = sqrt(sig);  
    
    % Stock path

    S(:,i+1)=S(:,i) .*(1 + r*dt + sigmaLV.*Z(:,i).*sqrt(dt));
    
    % We force at each time S(t)/M(t) to be a martingale

    S(:,i+1)= S(:,i+1) - mean(S(:,i+1)) + s0*exp(r*time(i));

    % Make sure that after moment matching we don't encounter negative stock values

    S(:,i+1)=max(S(:,i+1),0);

    % Adjust time

    time(i+1) = time(i) + dt;
end

% Compute implied volatility 

K=0.2:0.15:5;
IV_Hagan   = zeros(length(K),1);
CallPrices = zeros(length(K),1);
IV_MC      = zeros(length(K),1);
for i=1:length(K)
    IV_Hagan(i) =sigma(K(i),T);
    CallPrices(i)    =exp(-r*T).*mean(max(S(:,end)-K(i),0));
    IV_MC(i)    =ImpliedVolatility('c',CallPrices(i),K(i),T,s0,r,0.2); 
end

figure(1)
plot(K,IV_Hagan*100,'linewidth',1.5)
hold on
plot(K,IV_MC*100,'.r','MarkerSize',10)
title('Local Volatility model for European Options')
grid on;
xlabel('Strike, K')
ylabel('Implied volatility [%]')
plot(K,ones(length(K),1)*sigma_ATM*100,'--k','linewidth',1.5)
legend1 = legend('$$\hat\sigma(T,K)$$','$$\sigma_{imp}(T,K)$$','ATM Black-Scholes imp. vol');
set(legend1,'Interpreter','latex');

figure(2)

% Lognormal density (Black-Scholes) compared to density from generated paths

[y,x]=ksdensity(S(:,end),[0:0.01:7]);
f_S_logn =@(x)1./(sigma_ATM*x*sqrt(2*pi*T)).*exp(-(log(x./s0)-(r-0.5*sigma_ATM^2)*T).^2./(2*sigma_ATM*sigma_ATM*T));
plot(x,y,'linewidth',1.5)
hold on
plot(x,f_S_logn(x),'--r','linewidth',1.5)
xlabel('x')
ylabel('probability density function')
title('Local Volatility vs. Black Scholes')
grid on
legend('Local Volatility','Black Scholes')

% Implied volatility according to Hagan et al. SABR formula 

function impVol = HaganImpliedVolatility(K,T,f,alpha,beta,rho,gamma)
z        = gamma/alpha*(f*K).^((1-beta)/2).*log(f./K);
x_z      = log((sqrt(1-2*rho*z+z.^2)+z-rho)/(1-rho));
A        = alpha./((f*K).^((1-beta)/2).*(1+(1-beta)^2/24*log(f./K).^2+(1-beta)^4/1920*log(f./K).^4));
B1       = 1 + ((1-beta)^2/24*alpha^2./((f*K).^(1-beta))+1/4*(rho*beta*gamma*alpha)./((f*K).^((1-beta)/2))+(2-3*rho^2)/24*gamma^2)*T;
impVol   = A.*(z./x_z).*B1;

B2 = 1 + ((1-beta)^2/24*alpha^2./(f.^(2-2*beta))+1/4*(rho*beta*gamma*alpha)./(f.^(1-beta))+(2-3*rho^2)/24*gamma^2)*T;
impVol(K==f) = alpha/(f^(1-beta))*B2;

% Closed-form expression of European call/put option with Black-Scholes formula

function value=BS_Call_Option_Price(CP,S_0,K,sigma,tau,r)

% Black-Scholes call option price

d1    = (log(S_0 ./ K) + (r + 0.5 * sigma^2) * tau) / (sigma * sqrt(tau));
d2    = d1 - sigma * sqrt(tau);
if lower(CP) == 'c' || lower(CP) == 1
    value =normcdf(d1) * S_0 - normcdf(d2) .* K * exp(-r * tau);
elseif lower(CP) == 'p' || lower(CP) == -1
    value =normcdf(-d2) .* K*exp(-r*tau) - normcdf(-d1)*S_0;
end

function impliedVol = ImpliedVolatility(CP,marketPrice,K,T,S_0,r,initialVol)
    func = @(sigma) (BS_Call_Option_Price(CP,S_0,K,sigma,T,r) - marketPrice).^1.0;
    impliedVol = fzero(func,initialVol);
    
function value = LocalVarianceBasedOnSABR(s0,frwd,r,alpha,beta,rho,volvol)

% Define shock size for approximating derivatives

dt =0.001;
dx =0.001;

% Function hold for Hagan's implied volatility

sigma =@(x,t) HaganImpliedVolatility(x,t,frwd,alpha,beta,rho,volvol);

% Derivatives

dsigmadt   = @(x,t) (sigma(x,t+dt)-sigma(x,t))./dt;
dsigmadx   = @(x,t) (sigma(x+dx,t)-sigma(x-dx,t))./(2*dx);
d2sigmadx2 = @(x,t) (sigma(x+dx,t) + sigma(x-dx,t)-2*sigma(x,t))./(dx.*dx);

omega      = @(x,t) sigma(x,t).*sigma(x,t)*t;
domegadt   = @(x,t)sigma(x,t).^2 + 2*t*sigma(x,t).*dsigmadt(x,t);
domegadx   = @(x,t)2*t*sigma(x,t).*dsigmadx(x,t);
d2omegadx2 = @(x,t)2*t*(dsigmadx(x,t)).^2 + 2*t*sigma(x,t).*d2sigmadx2(x,t);

term1    = @(x,t)1+x.*domegadx(x,t).*(0.5 - log(x./(s0*exp(r*t)))./omega(x,t));
term2    = @(x,t)0.5*x.^2.*d2omegadx2(x,t);
term3    = @(x,t)0.5*x.^2.*domegadx(x,t).^2.*(-1/8-1./(2.*omega(x,t))+log(x./(s0*exp(r*t))).*log(x./(s0*exp(r*t)))./(2*omega(x,t).*omega(x,t)));

% Final expression for local variance

sigmalv2 = @(x,t)(domegadt(x,t)+r*x.*domegadx(x,t))./(term1(x,t)+term2(x,t)+term3(x,t));
value = sigmalv2;

File: PythonCodes/Chapter 04/Fig04_10.py
#%%
"""
Created on Feb 11 2019
Local volatility model based on the implied volatility obtained from Hagan's formula
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import scipy.optimize as optimize
import enum 

# This class defines puts and calls

class OptionType(enum.Enum):
    CALL = 1.0
    PUT = -1.0

def HaganImpliedVolatility(K,T,f,alpha,beta,rho,gamma):

    # We make sure that the input is of array type

    if type(K) == float:
        K = np.array([K])
    if K is not np.array:
        K = np.array(K).reshape([len(K),1])
    
    # The strike prices cannot be too close to 0

    K[K<1e-10] = 1e-10
    
    z        = gamma/alpha*np.power(f*K,(1.0-beta)/2.0)*np.log(f/K)
    x_z      = np.log((np.sqrt(1.0-2.0*rho*z+z*z)+z-rho)/(1.0-rho))
    A        = alpha/(np.power(f*K,((1.0-beta)/2.0))*(1.0+np.power(1.0-beta,2.0)/24.0*
                               np.power(np.log(f/K),2.0)+np.power((1.0-beta),4.0)/1920.0*
                               np.power(np.log(f/K),4.0)))
    B1       = 1.0 + (np.power((1.0-beta),2.0)/24.0*alpha*alpha/(np.power((f*K),
                1-beta))+1/4*(rho*beta*gamma*alpha)/(np.power((f*K),
                             ((1.0-beta)/2.0)))+(2.0-3.0*rho*rho)/24.0*gamma*gamma)*T
    impVol   = A*(z/x_z) * B1
    B2 = 1.0 + (np.power(1.0-beta,2.0)/24.0*alpha*alpha/
                (np.power(f,2.0-2.0*beta))+1.0/4.0*(rho*beta*gamma*
                alpha)/np.power(f,(1.0-beta))+(2.0-3.0*rho*rho)/24.0*gamma*gamma)*T

    # Special treatment of ATM strike value

    impVol[np.where(K==f)] = alpha / np.power(f,(1-beta)) * B2
    return impVol

def LocalVarianceBasedOnSABR(s0,frwd,r,alpha,beta,rho,volvol):

    # Define shock size for approximating derivatives

    dt =0.001
    dx =0.001

    # Function for Hagan's implied volatility approximation 

    sigma =lambda x,t: HaganImpliedVolatility(x,t,frwd,alpha,beta,rho,volvol)

    # Derivatives

    dsigmadt   = lambda x,t: (sigma(x,t+dt)-sigma(x,t))/dt
    dsigmadx   = lambda x,t: (sigma(x+dx,t)-sigma(x-dx,t))/(2.0*dx)
    d2sigmadx2 = lambda x,t: (sigma(x+dx,t) + sigma(x-dx,t)-2.0*sigma(x,t))/(dx*dx)
    omega      = lambda x,t: sigma(x,t)*sigma(x,t)*t
    domegadt   = lambda x,t: sigma(x,t)**2.0 + 2.0*t*sigma(x,t)*dsigmadt(x,t)
    domegadx   = lambda x,t: 2.0*t*sigma(x,t)*dsigmadx(x,t)
    #d2omegadx2 = lambda x,t: 2.0*t*(dsigmadx(x,t))**2.0 + 2.0*t*sigma(x,t)*d2sigmadx2(x,t)
    d2omegadx2 = lambda x,t: 2.0*t*np.power(dsigmadx(x,t),2.0) + 2.0*t*sigma(x,t)*d2sigmadx2(x,t)

    term1    = lambda x,t:  1.0+x*domegadx(x,t)*(0.5 - np.log(x/(s0*np.exp(r*t))) / omega(x,t))    
    term2    = lambda x,t:  0.5*np.power(x,2.0)*d2omegadx2(x,t)
    term3    = lambda x,t:  0.5*np.power(x,2.0)*np.power(domegadx(x,t),2.0)*(-1.0/8.0-1.0/(2.0*omega(x,t))\
            +np.log(x/(s0*np.exp(r*t)))*np.log(x/(s0*np.exp(r*t)))/(2*omega(x,t)*omega(x,t)))

    # Final expression for local variance

    sigmalv2 = lambda x,t:(domegadt(x,t)+r*x*domegadx(x,t))/(term1(x,t)+term2(x,t)+term3(x,t))
    return sigmalv2
 
    # Black-Scholes call option price

def BS_Call_Put_Option_Price(CP,S_0,K,sigma,tau,r):
    if K is list:
        K = np.array(K).reshape([len(K),1])
    d1    = (np.log(S_0 / K) + (r + 0.5 * np.power(sigma,2.0)) 
    * tau) / (sigma * np.sqrt(tau))
    d2    = d1 - sigma * np.sqrt(tau)
    if CP == OptionType.CALL:
        value = st.norm.cdf(d1) * S_0 - st.norm.cdf(d2) * K * np.exp(-r * tau)
    elif CP == OptionType.PUT:
        value = st.norm.cdf(-d2) * K * np.exp(-r * tau) - st.norm.cdf(-d1)*S_0
    return value

# Implied volatility method

def ImpliedVolatility(CP,marketPrice,K,T,S_0,r):

    # To determine initial volatility we define a grid for sigma
    # and interpolate on the inverse function

    sigmaGrid = np.linspace(0,2,200)
    optPriceGrid = BS_Call_Put_Option_Price(CP,S_0,K,sigmaGrid,T,r)
    sigmaInitial = np.interp(marketPrice,optPriceGrid,sigmaGrid)
    print("Initial volatility = {0}".format(sigmaInitial))
    
    # Use already determined input for the local-search (final tuning)

    func = lambda sigma: np.power(BS_Call_Put_Option_Price(CP,S_0,K,sigma,T,r) - marketPrice, 1.0)
    impliedVol = optimize.newton(func, sigmaInitial, tol=1e-15)
    print("Final volatility = {0}".format(impliedVol))
    return impliedVol

def EUOptionPriceFromMCPaths(CP,S,K,T,r):

    # S is a vector of Monte Carlo samples at T

    if CP == OptionType.CALL:
        return np.exp(-r*T)*np.mean(np.maximum(S-K,0.0))
    elif CP == OptionType.PUT:
        return np.exp(-r*T)*np.mean(np.maximum(K-S,0.0))
    
def mainCalculation():

    # For the SABR model we take beta =1 and rho =0 (as simplification)

    beta   = 1.0
    rho    = 0.0

    # Other model parameters

    volvol = 0.2
    s0     = 1.0
    T      = 10.0
    r      = 0.05
    alpha  = 0.2
    f_0    = s0*np.exp(r*T)
    CP     = OptionType.CALL

    # Monte Carlo settings

    NoOfPaths = 25000
    NoOfSteps = (int)(100*T)
    
    # We define the market to be driven by Hagan's SABR formula
    # Based on this formula we derive the local volatility/variance 

    sigma     = lambda x,t: HaganImpliedVolatility(x,t,f_0,alpha,beta,rho,volvol)
    
    # Local variance based on the Hagan's SABR formula

    sigmalv2  = LocalVarianceBasedOnSABR(s0,f_0,r,alpha,beta,rho,volvol)
           
    # Monte Carlo simulation

    dt        = T/NoOfSteps
    np.random.seed(4)
    Z         = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    S         = np.zeros([NoOfPaths,NoOfSteps+1])
    
    S[:,0]    = s0;
    time = np.zeros([NoOfSteps+1,1])
    
    for i in range(0,NoOfSteps):

        # This condition is necessary as for t=0 we cannot compute implied
        # volatilities

        if time[i]==0.0:
            time[i]=0.0001
        
        print('current time is {0}'.format(time[i]))

        # Standarize Normal(0,1)

        Z[:,i]=(Z[:,i]-np.mean(Z[:,i]))/np.std(Z[:,i])
        
        # Compute local volatility

        S_i = np.array(S[:,i]).reshape([NoOfPaths,1])
        temp = sigmalv2(S_i,time[i])
        sig = np.real(temp)
        np.nan_to_num(sig)
        
        # Because of discretizations we may encouter negative variance which
        # is set to 0 here.

        sig=np.maximum(sig,1e-14)
        sigmaLV = np.sqrt(sig)  
        
        # Stock path

        S[:,i+1]=S[:,i] * (1.0 + r*dt + sigmaLV.transpose()*Z[:,i]*np.sqrt(dt))
                       
        # We force that at each time S(t)/M(t) is a martingale

        S[:,i+1]= S[:,i+1] - np.mean(S[:,i+1]) + s0*np.exp(r*time[i])
        
        # Make sure that after moment matching we don't encounter negative stock values

        S[:,i+1]=np.maximum(S[:,i+1],1e-14)
        
        # Adjust time

        time[i+1] = time[i] + dt

    # Plot some results

    K = np.linspace(0.2,5.0,25)
    #c_n = np.array([-1.5, -1.0, -0.5,0.0, 0.5, 1.0, 1.5])
    #K= s0*np.exp(r*T) * np.exp(0.1 * c_n * np.sqrt(T))
    OptPrice = np.zeros([len(K),1])
    IV_Hagan = np.zeros([len(K),1])
    IV_MC = np.zeros([len(K),1])
    for (idx,k) in enumerate(K):
        OptPrice[idx] = EUOptionPriceFromMCPaths(CP,S[:,-1],k,T,r)
        IV_Hagan[idx] = sigma([k],T)*100.0
        IV_MC[idx]    = ImpliedVolatility(CP,OptPrice[idx],k,T,s0,r)*100.0
        
    # Plot the option prices

    plt.figure(1)
    plt.plot(K,OptPrice)
    plt.grid()
    plt.xlabel('strike')
    plt.ylabel('option price')
    
    # Plot the implied volatilities

    plt.figure(2)
    plt.plot(K,IV_Hagan)
    plt.plot(K,IV_MC,'-r')
    plt.grid()
    plt.xlabel('strike')
    plt.ylabel('implied volatility')
    plt.legend(['Hagan','Monte Carlo'])
    plt.axis([np.min(K),np.max(K),0,40])
mainCalculation()
